chainer/function.py

```diff
@@ -55,111 +55,138 @@ def force_backprop_mode():
+    @property
+    def label(self):
+        return self._function.label
 
-                    |--- f <--- y.node
-           x.node <-+
-                    |--- g <--- z.node
+    @property
+    def _impl_name(self):
+        return self._function.__class__.__name__
 
-       Note that the branching is correctly managed on backward computation,
-       i.e. the gradients from ``f`` and ``g`` are accumulated to the gradient
-       of ``x``.
+    def check_type_forward(self, in_types):
+        self._function.check_type_forward(in_types)
 
-    Every function implementation should provide :meth:`forward_cpu`,
-    :meth:`forward_gpu`, :meth:`backward_cpu` and :meth:`backward_gpu`.
-    Alternatively, one can provide :meth:`forward` and :meth:`backward` instead
-    of separate methods. Backward methods have default implementations that
-    just return ``None``, which indicates that the function is non-
-    differentiable.
+    def forward(self, inputs):
+        # Retain all inputs by default in old-style functions.
+        self.retain_inputs(six.moves.range(len(inputs)))
+        return self._function.forward(inputs)
+
+    def backward(self, target_input_indexes, grad_outputs):
+        in_data = tuple([input.data for input in self.inputs])
+        grad_out_data = tuple([None if grad is None else grad.data
+                               for grad in grad_outputs])
+
+        with cuda.get_device_from_array(*(in_data + grad_out_data)):
+            gxs = self._function.backward(in_data, grad_out_data)
+        for x, gx in six.moves.zip(self.inputs, gxs):
+            variable._check_grad_type(self, x, gx)
+
+        ret = []
+        for i in target_input_indexes:
+            if gxs[i] is None:
+                g = None
+            else:
+                # Intentionallly not passing requires_grad=False so that
+                # backprop routines can raise an error when a further backprop
+                # is attempted against this gradient variable.
+                g = variable.Variable(gxs[i])
+                g.node._old_style_grad_generator = self._function.label
+            ret.append(g)
+
+        return tuple(ret)
 
-    For functions that do not need a part of inputs in backward computation,
-    there is a way to possibly reduce the memory consumption by quickly
-    releasing the input arrays after the forward propagation. This is done by
-    calling :meth:`retain_inputs` from inside of :meth:`forward` (including
-    :meth:`forward_cpu` and :meth:`forward_gpu`). See the documentation of
-    :meth:`retain_inputs` for details.
 
-    For functions that need a part of outputs in backward computation, it is
-    **strongly recommended** to call :meth:`retain_outputs` from inside of
-    :meth:`forward` (including :meth:`forward_cpu` and :meth:`forward_gpu`).
-    It marks the specified output variable nodes to retain the data. The
-    retained data can be accessed by :attr:`output_arrays` property.
+class Function(object):
```

Comment:
Are you thinking of removing Function class in the future major version? Although it is not recommended to use this API, its removal seems to affect virtually all existing user code.